Page tree ajax request to fetch children

melegiul opened the pull request:
### Short description
<!-- Describe this PR in one or two sentences. -->
Fetch page tree children by ajax request

### Proposed changes
<!-- Describe this PR in more detail. -->

- Page tree starts with root nodes only
- On expanding new page children get queried and added to the DOM

### Resolved issues
<!-- List all issues which should be closed when this PR is merged. -->

Fixes: #1101

timoludwig reviewed (commented):
Amazing, thanks a lot! :muscle: :rocket: :heart_eyes: 
I noticed a few problems, have a look at my comments below for details.

Apart from that, I think it would be cool to load the next children in advance after the page load and before the sub-pages are expanded, so that when the user clicks on expand, the results are already there... but this is probably out of scope for this PR, let's open another issue after this is finished...

melegiul commented on integreat_cms/cms/templates/pages/page_tree_node.html:
```diff
@@ -4,8 +4,8 @@
 {% get_translation page language.slug as page_translation %}
-<tr id="page-{{ page.id }}-drop-left" data-drop-id="{{ page.id }}" data-drop-position="left" class="drop {% if page.relative_depth == 1 %} drop-between {% endif %} h-3 -m-3 hidden level-{{page.relative_depth}}"><td colspan="9"><div><span></span></div></td></tr>
-<tr id="page-{{ page.id }}" data-drop-id="{{ page.id }}" data-drop-position="last-child" class="drop drop-on border-t border-b border-solid border-gray-200 hover:bg-gray-100 {% if page.relative_depth > 1 %} level-{{page.relative_depth}} {% if not filter_form.has_changed %} hidden {% endif %} {% endif %}">
+<tr id="page-{{ page.id }}-drop-left" data-drop-id="{{ page.id }}" data-drop-position="left" class="drop drop-between h-3 -m-3 hidden level-{{page.relative_depth}}"><td colspan="9"><div><span></span></div></td></tr>
```
Thank you very much for your review, I applied your suggestions so far and they work greatly.
Here I think we have a problem, committing this suggestion will break drag and drop, because the relative depth for children of roots is 2, so they will not render this additional drop region above the current page after expanding, I remove this if statement therefore.
But I can try to find an if statement, so that it will render for children and not for search results and archived pages

melegiul commented on integreat_cms/static/src/js/pages/collapse-subpages.ts:
```diff
@@ -65,8 +76,44 @@ function toggleSubpagesRecursive(collapseSpan: HTMLElement) {
+    currentRow = rowToinsert;
+  });
+  // render rows buttton icons
+  feather.replace({ class: 'inline-block' });
```
Thanks for reporting this, in 902aeb5b8d86d9a1305b59cbdbe5420c5c767e26 I have added the listeners and also set the confirmation context by creating a new constant, which is used now in the view and in the mixin
Now the confirmation dialogs should work

melegiul commented on integreat_cms/static/src/js/pages/collapse-subpages.ts:
```diff
@@ -14,12 +16,25 @@ window.addEventListener("load", () => {
+    .catch(e => {
+      console.log('There has been a problem with your fetch operation: ' + e.message);
+    });
+    addDragAndDropListeners();
```
e9dcf9c50e259166e259e231b12a29534b0a6a8d this will iterate over all ancestors and add the recently added children to their list of descendants on children expansion

timoludwig commented on integreat_cms/cms/views/pages/page_actions.py:
```diff
@@ -928,3 +929,70 @@ def render_mirrored_page_field(request, region_slug):
+    parent = pages[depth - 1]
+    # The remaining pages are the direct children
+    children = pages[depth:]
+    ancestors = [ancestor.id for ancestor in pages[:depth]]
```
```suggestion
    # For every depth level, exactly one ancestor is in the list
    ancestor_ids = [ancestor.id for ancestor in pages[:depth]]
```
I think it's better to make it clear what this list only contains ids and not the objects itself.

melegiul commented on integreat_cms/cms/templates/pages/page_tree_node.html:
```diff
@@ -18,13 +18,14 @@
                 <i data-feather="move"></i>
             </span>
-            {% if page.cached_children|length > 0 %}
+            {% if not page.is_leaf %}
```
What you think about this solution d280d82955ee5ad9a307c3207ed5761e53c3e4ff?
![expansion](https://user-images.githubusercontent.com/47010475/153048845-7760604c-a54e-4f46-8022-ac7369677a5e.gif)

timoludwig replied:
To be honest, I'm not 100% convinced, but I think as an intermediary solution, this is fine. Probably this only happens in rare edge cases, so shouldn't be a big problem. In the long term, we can maybe start the fetching of the subpages before the user clicks on the button, so we can remove the expansion button dynamically where the subpages don't exist.

timoludwig commented on integreat_cms/cms/templates/messages.html:
```diff
@@ -24,3 +26,6 @@
     {% endfor %}
 </div>
 {% endif %}
+<div id="archived-subpages-only" class="bg-blue-100 border-l-4 border-blue-500 text-blue-500 px-4 py-3 mb-5 hidden" role="alert">
+    <p>{% trans 'All subpages have been archived' %}</p>
+</div>
```
Since the message template is embedded directly before the normal content, I think we should place this div at the beginning of the content of the page tree view template instead of the end of the messages template...
https://github.com/digitalfabrik/integreat-cms/blob/e76ad67d2e165702acafb089aec676d9203a9b56/integreat_cms/cms/templates/_base.html#L265-L266
Also, I'd probably change the message to "All subpages of this page are archived".

timoludwig reviewed (commented):
Thanks for your changes! Only a few details are missing now.
And could you add a small message to the changelog about this?

melegiul commented:
@timoludwig Thanks for your hints, I have now added a changelog entry, moved the archived message box and switched to a class based view
But there still remain probelms at the archived page tree view and with page filtering
I will have a look at them

melegiul commented:
dab25ce22301e04de593f206a8f884a14af73f15 this should fix the broken archived page tree
Page filters seem to be fine too

timoludwig commented on integreat_cms/cms/views/pages/partial_page_tree_view.py:
```diff
@@ -0,0 +1,78 @@
+        # List of all ancestors, the requested parent and all direct children
+        pages = region.pages.filter(
+            # This condition queries the parent and all direct children
+            Q(tree_id=tree_id, lft__range=(lft, int(rgt) - 1), depth__lte=depth + 1)
```
```suggestion
            Q(tree_id=tree_id, lft__range=(lft, rgt - 1), depth__lte=depth + 1)
```
Since you're casting `rgt` to int anyway, it's not required to do it again.

timoludwig commented on integreat_cms/locale/de/LC_MESSAGES/django.po:
```diff
@@ -4557,12 +4557,16 @@ msgstr ""
-#: cms/templates/pages/page_tree.html:13
+#: cms/templates/pages/page_tree.html:12
+msgid "All subpages of this page are archived"
+msgstr "Alle Unterseiten von dieser Seite sind archiviert"
```
```suggestion
msgstr "Alle Unterseiten dieser Seite sind archiviert"
```

timoludwig commented on integreat_cms/static/src/js/pages/collapse-subpages.ts:
```diff
@@ -65,8 +91,91 @@ function toggleSubpagesRecursive(collapseSpan: HTMLElement) {
+      ancestorDragSpan.dataset.nodeDescendants
+    ) as number[];
+    // update descendants
+    ancestorDragSpan.dataset.nodeDescendants = "[" + descendants.concat(childrenIds).join(",") + "]";
```
```suggestion
    ancestorDragSpan.dataset.nodeDescendants = `[${descendants.concat(childrenIds)}]`;
```
Lists are automatically joined with commas when converted to a string.

timoludwig commented on integreat_cms/static/src/js/pages/collapse-subpages.ts:
```diff
@@ -65,8 +91,91 @@ function toggleSubpagesRecursive(collapseSpan: HTMLElement) {
+  const currentChidlrenIds = JSON.parse(
+    collapseSpan.dataset.pageChildren
+  ) as number[];
+  collapseSpan.dataset.pageChildren = "[" + currentChidlrenIds.concat(childrenIds).join(",") + "]";
```
```suggestion
  collapseSpan.dataset.pageChildren = `[${currentChidlrenIds.concat(childrenIds)}]`;
```

timoludwig reviewed (commented):
Awesome, thanks! :rocket: 
Just a few tiny details, then this PR is ready to merge.
Could you rebase your branch to resolve the conflicts and clean up the git history a bit?

timoludwig reviewed (approved):
Amazing, thanks! :pray:
